In [1]:
# Cargar el archivo 
import pandas as pd
import numpy as np
url = "https://raw.githubusercontent.com/mricardo89/data-mining/main/Unit-II/datasets/googleplaystore.csv"
df = pd.read_csv(url)

# Para el reto 1
def size_parser(s:str):
    if s == 'Varies with device':
        return np.nan
    elif s.endswith('M'):
        return float(s.replace('M', ''))
    elif s.endswith('k'):
        return float(s.replace('k', '')) / 1024
    else:
        return np.nan

df['Size'] = df['Size'].apply(size_parser)
print("Apps avg size (M):", df['Size'].mean())

Apps avg size (M): 21.51616543577433


**Reto 1: La Trampa de las Unidades de Medida**

El equipo de infraestructura quiere saber si las aplicaciones más pesadas tienen menos descargas. Para averiguarlo, necesitan que la columna Size sea completamente numérica. Sin embargo, si revisan la columna, encontrarán valores como "19M" (Megabytes), "201k" (Kilobytes) y un texto molesto que dice "Varies with device".
- Tu misión algorítmica:
    - Escribe una función pura en Python que reciba un string. Si el string termina en 'M', quítale la 'M' y conviértelo a flotante (dejándolo como Megabytes). Si termina en 'k', quítale la 'k', conviértelo a flotante y divídelo entre 1024 (para pasarlo también a Megabytes). Si dice "Varies with device", devuélvelo como un valor nulo de Numpy (np.nan).
    - Aplica esta función a toda la columna utilizando el método .apply().
- Pregunta a responder: Una vez convertida la columna a valores numéricos (Megabytes), ejecuta el método .mean(). ¿Cuál es el peso promedio en Megabytes de las apps en la Play Store?

El peso en promedoio de las apps es de 21.5161 Megabytes.

In [3]:
# Reto 2
df['Last Updated'] = pd.to_datetime(df['Last Updated'], errors='coerce')
df['Year Updated'] = df['Last Updated'].dt.year
print(df['Year Updated'].value_counts())

Year Updated
2018.0    7349
2017.0    1867
2016.0     804
2015.0     459
2014.0     209
2013.0     110
2012.0      26
2011.0      15
2010.0       1
Name: count, dtype: int64


**Reto 2: El Tipo de Dato Cronológico**

Ningún análisis de software está completo sin analizar el tiempo. La columna Last Updated tiene fechas escritas como texto: "January 7, 2018". Para un modelo matemático o una serie de tiempo, eso es texto inservible.
- Tu misión algorítmica:
    - Investiga y utiliza la función pd.to_datetime() de Pandas para sobrescribir la columna Last Updated, convirtiéndola del tipo string (Object) al tipo nativo datetime64.   
    - Ahora que es un objeto de tiempo, Pandas te permite extraer componentes específicos. Crea una nueva columna llamada Year_Updated extrayendo únicamente el año (df['Last Updated'].dt.year).
- Pregunta a responder: Utilizando la sumarización categórica (value_counts()) sobre tu nueva columna Year_Updated, ¿en qué año se actualizó la mayor cantidad de aplicaciones en nuestro dataset?

Fue en año 2018.

In [5]:
df['Size'].isna().value_counts()

Size
False    9145
True     1696
Name: count, dtype: int64

**Reto 3: La Decisión Arquitectónica**
Al resolver el Reto 1, introdujiste intencionalmente valores NaN en las aplicaciones cuyo tamaño decía "Varies with device".
- Tu misión algorítmica: Evalúa cuántos registros quedaron vacíos. Como ingenieros, decidan y apliquen la mejor técnica: ¿es matemáticamente más sano borrar esas filas con un .dropna() porque el peso de la app es crítico, o prefieren rellenar ese hueco imputando la mediana global del peso de las apps?
- Pregunta a responder: Redacten una breve justificación técnica de 3 líneas explicando qué método eligieron y por qué lo consideran superior para no dañar al modelo de predicción.

Si se eliminaran esos registro estaríamos peridendo 1696 entradas. Es mejor sustituir los valors NaN con la mediana ya que así evitamos la pérdida de datos y la mediana es menos sensible a valores extremos que la media. Si los eliminaramos podriamos perder otra información valiosa y afectar a otras columnas, lo que llevaría que al final afectaramos nuestro modelo. 